library/data/registry-entry-attributes/src/main/java/org/quiltmc/qsl/registry/attribute/impl/RegistryEntryAttributeReloader.java

```diff
@@ -0,0 +1,225 @@
+				profiler.push(ID + "/finding_resources/" + path);
+
+				Collection<Identifier> jsonIds = manager.findResources("attributes/" + path, s -> s.endsWith(".json"));
+				if (jsonIds.isEmpty()) {
+					continue;
+				}
+
+				Registry<?> registry = entry.getValue();
+				processResources(manager, profiler, attributeMaps, jsonIds, registry);
+
+				profiler.pop();
+			}
+
+			return new LoadedData(attributeMaps);
+		}, executor);
+	}
+
+	private void processResources(ResourceManager manager, Profiler profiler,
+								  Map<RegistryEntryAttribute<?, ?>, AttributeMap> attributeMaps,
+								  Collection<Identifier> jsonIds, Registry<?> registry) {
+		for (var jsonId : jsonIds) {
+			Identifier attribId = getAttributeId(jsonId);
+			RegistryEntryAttribute<?, ?> attrib = RegistryEntryAttributeHolderImpl.getAttribute(registry, attribId);
+			if (attrib == null) {
+				LOGGER.warn("Unknown attribute {} (from {})", attribId, jsonId);
+				continue;
+			}
+
+			profiler.swap(ID + "/getting_resources{" + jsonId + "}");
+
+			List<Resource> resources;
+			try {
+				resources = manager.getAllResources(jsonId);
+			} catch (IOException e) {
+				LOGGER.error("Failed to get all resources for {}", jsonId);
+				LOGGER.catching(Level.ERROR, e);
+				continue;
+			}
+
+			profiler.swap(ID + "/processing_resources{" + jsonId + "," + attribId + "}");
+
+			AttributeMap attribMap = attributeMaps.computeIfAbsent(attrib,
+					id -> new AttributeMap(registry, attrib));
+			for (var resource : resources) {
+				attribMap.processResource(resource);
+			}
+		}
+	}
+
+	@Override
+	public CompletableFuture<Void> apply(LoadedData data, ResourceManager manager, Profiler profiler, Executor executor) {
+		return CompletableFuture.runAsync(() -> data.apply(profiler), executor);
+	}
+
+	// "<namespace>:attributes/<path>/<file_name>.json" becomes "<namespace>:<file_name>"
+	private Identifier getAttributeId(Identifier jsonId) {
+		String path = jsonId.getPath();
+		int lastSlash = path.lastIndexOf('/');
+		path = path.substring(lastSlash + 1);
+
+		int lastDot = path.lastIndexOf('.');
+		path = path.substring(0, lastDot);
+		return new Identifier(jsonId.getNamespace(), path);
+	}
+
+	protected record LoadedData(Map<RegistryEntryAttribute<?, ?>, AttributeMap> attributeMaps) {
+		public void apply(Profiler profiler) {
+			profiler.push(ID + "/clear_attributes");
+
+			for (var entry : Registry.REGISTRIES.getEntries()) {
+				RegistryEntryAttributeHolderImpl.getData(entry.getValue()).clear();
+			}
+
+			for (Map.Entry<RegistryEntryAttribute<?, ?>, AttributeMap> entry : attributeMaps.entrySet()) {
+				profiler.swap(ID + "/apply_attribute{" + entry.getKey().getId() + "}");
+				applyOne(entry.getKey(), entry.getValue());
+			}
+
+			profiler.pop();
+		}
+
+		@SuppressWarnings("unchecked")
+		private <R, V> void applyOne(RegistryEntryAttribute<R, V> attrib, AttributeMap attribMap) {
+			var registry = attrib.getRegistry();
+			assert registry != null : "bruh momento";
```

Comment:
Maybe use `Objects.requireNonNull` instead of an assert?
~~also is this assert even necessary?~~
```suggestion
			Objects.requireNonNull(registry, "registry");
```

Reply:
Assert is something you would use in a dev environment to ensure some value makes sense. For failing at runtime requireNonNull is more appropriate.